src/baldr/graphreader.cc

```diff
@@ -12,9 +12,220 @@
 #include "baldr/curl_tilegetter.h"
 #include "filesystem.h"
 
+// move these and the stuff below
+#include "midgard/sequence.h"
+#include <chrono>
+#include <condition_variable>
+#include <ctime>
+#include <memory>
+#include <mutex>
+#include <thread>
+
 using namespace valhalla::midgard;
 
 namespace {
+#ifdef _MSC_VER
+#define MTIME(st_stat) st_stat.st_mtime
+#elif __APPLE__
+#define MTIME(st_stat) st_stat.st_mtime
+#else
+#define MTIME(st_stat) st_stat.st_mtim.tv_sec
+#endif
+
+// some typedefs until we have the real things
+using incident_tile_t = std::vector<char>;
+using incident_cache_t = std::unordered_map<uint64_t, std::shared_ptr<incident_tile_t>>;
+
+/**
+ * Read the contents of a file into an incident tile
+ * @param filename   name of the file on the file system to read into memory
+ * @return a shared pointer with the data of the tile or an empty pointer if it could not be read
+ */
+std::shared_ptr<incident_tile_t> read_tile(const std::string& filename) {
+  // crack the file open
+  std::ifstream file(filename, std::ios::in | std::ios::binary | std::ios::ate);
+  std::shared_ptr<incident_tile_t> tile;
+  if (file.is_open()) {
+    // get the size and allocate it
+    size_t filesize = file.tellg();
+    tile.reset(new std::vector<char>(filesize));
+    // read the data in
+    file.seekg(0, std::ios::beg);
+    file.read(tile->data(), filesize);
+    file.close();
+  }
+  return tile;
+}
+
+/**
+ * Thread work function that continually checks for updates to incident tiles. The updates come in
+ * the form of timestamps in a memory mapped file of 64bit integers (one per tile). The first 25 bits
+ * are the tile id and level (as a normal graphid) the remaining 39 bits are the timestamp. When
+ * the timestamp for the last check is older than a timestamp for a given file that file is replaced
+ * with whatever its contents are on disk.
```

Comment:
Perhaps an additional bit or two to describe which buffer in a double buffer contains the newest update?

Reply:
the temporary implementation here didnt actually end up using a double buffer approach. instead it does asynchronous loading and uses shared ptrs to transfer ownership to the reader when the writer invalidates a tile in its store